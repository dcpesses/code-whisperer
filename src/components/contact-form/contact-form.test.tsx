import {render, screen} from '@testing-library/react';
import {UserEvent, userEvent} from '@testing-library/user-event';
import {Mock} from 'vitest';
import ContactForm from './index';

describe('ContactForm', () => {

  let onFormSubmitSpy: Mock;
  let setValidated: React.Dispatch<React.SetStateAction<boolean>>;
  let user: UserEvent;

  beforeEach(()=>{
    onFormSubmitSpy = vi.fn();
    setValidated = vi.fn();
    user = userEvent.setup();
  });

  const onFormSubmit = (event: SubmitEvent) => {
    const form:HTMLFormElement = event.currentTarget as HTMLFormElement;

    // spy on data via mock.calls
    const formData = new FormData(form);
    const formDataEntries = [...formData.entries()]
      .reduce((acc: Record<string, FormDataEntryValue>, [key, value]) => {
        acc[key] = value;
        return acc;
      }, {}
      );
    onFormSubmitSpy(formDataEntries);

    // actual logic
    if (form.checkValidity() === false) {
      event.preventDefault();
      event.stopPropagation();
    } else {
      setValidated(true);
    }
  };

  test('Should render as expected', () => {
    render(<ContactForm />);

    const form = screen.getByTestId('contact-form');
    expect(form).toMatchSnapshot();
    expect(screen.getByText('Submit')).toBeTruthy();
  });

  test('should submit when form has valid data', async() => {
    render(<ContactForm disabled={false} onSubmit={onFormSubmit} validated={true} />);

    const form = screen.getByTestId('contact-form');
    expect(form).toMatchSnapshot('pristine');

    const nameInput:HTMLInputElement = screen.getByLabelText('Full Name *');
    const emailInput:HTMLInputElement = screen.getByLabelText('Email Address *');
    const commentsInput:HTMLInputElement = screen.getByLabelText('Comments/Questions *');
    const btnSubmit:HTMLButtonElement = screen.getByRole('button', { name: 'Submit' });

    await user.type(nameInput, 'John Doe');
    await user.type(emailInput, 'john.doe@example.com');
    await user.type(commentsInput, 'I love this app!');

    await user.click(btnSubmit);

    expect(onFormSubmitSpy).toBeCalled();
    expect(onFormSubmitSpy.mock.calls[0][0]).toMatchObject(
      {
        'Comments/Questions': 'I love this app!',
        'Email Address': 'john.doe@example.com',
        'Full Name': 'John Doe',
        'g-recaptcha-response': '',
        'subscribe_a7fdfc1ea41e_48062': '',
      }
    );
    expect(setValidated).toBeCalled();
    expect(form.classList).toContain('was-validated');
    expect(form).toMatchSnapshot('dirty');
  });
  test('should not submit when form has invalid data', async() => {
    const {rerender} = render(<ContactForm disabled={false} onSubmit={onFormSubmit} validated={false} />);

    const form = screen.getByTestId('contact-form');
    expect(form).toMatchSnapshot('pristine');

    const nameInput:HTMLInputElement = screen.getByLabelText('Full Name *');
    const emailInput:HTMLInputElement = screen.getByLabelText('Email Address *');
    const btnSubmit:HTMLButtonElement = screen.getByRole('button', { name: 'Submit' });

    await user.type(nameInput, 'John Doe');
    await user.type(emailInput, 'john.doe');

    await user.click(btnSubmit);

    rerender(<ContactForm disabled={false} onSubmit={onFormSubmit} validated={true} />);

    expect(onFormSubmitSpy).toBeCalled();
    expect(setValidated).not.toBeCalled();

    expect(onFormSubmitSpy.mock.calls[0][0]).toMatchObject(
      {
        'Comments/Questions': '',
        'Email Address': 'john.doe',
        'Full Name': 'John Doe',
        'g-recaptcha-response': '',
        'subscribe_a7fdfc1ea41e_48062': '',
      }
    );

    expect(nameInput.reportValidity()).toBeTruthy();
    expect(emailInput.reportValidity()).toBeFalsy();

    expect(form.classList).toContain('was-validated');
    expect(form).toMatchSnapshot('dirty');
  });
  test('should not submit when disabled', async() => {
    const {rerender} = render(<ContactForm disabled={true} onSubmit={onFormSubmit} validated={false} />);

    const form = screen.getByTestId('contact-form');
    expect(form).toMatchSnapshot('pristine');

    const btnSubmit = screen.getByRole('button', { name: 'Submit' });

    await user.click(btnSubmit);

    rerender(<ContactForm disabled={false} onSubmit={onFormSubmit} validated={true} />);

    expect(onFormSubmitSpy).not.toBeCalled();
    expect(form).toMatchSnapshot('dirty');
  });
});
