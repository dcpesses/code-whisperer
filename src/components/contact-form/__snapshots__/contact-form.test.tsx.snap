// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContactForm > Should render as expected 1`] = `
<form
  accept-charset="UTF-8"
  action="/submit"
  class=""
  data-testid="contact-form"
  enctype=""
  id="contact-form"
  method="GET"
  novalidate=""
>
  <div>
    <label
      class="form-label"
      for="contact.Full_Name"
    >
      Full Name *
    </label>
    <input
      class="form-control"
      id="contact.Full_Name"
      name="Full Name"
      placeholder="First name"
      required=""
      type="text"
    />
    <div
      class="invalid-feedback"
    >
      Please enter your full name
    </div>
  </div>
  <div>
    <label
      class="form-label"
      for="contact.Email_Address"
    >
      Email Address *
    </label>
    <input
      class="form-control"
      id="contact.Email_Address"
      name="Email Address"
      required=""
      type="email"
    />
    <div
      class="invalid-feedback"
    >
      Please enter a valid email address
    </div>
  </div>
  <div>
    <label
      class="form-label"
      for="contact.Comments_Questions"
    >
      Comments/Questions *
    </label>
    <textarea
      class="form-control"
      id="contact.Comments_Questions"
      name="Comments/Questions"
      required=""
    />
    <div
      class="invalid-feedback"
    >
      Please enter your feedback.
    </div>
  </div>
  <div
    style="opacity: 0; position: absolute; top: 0px; left: -5000px; height: 0px; width: 0px;"
  >
    <label
      for="subscribe_a7fdfc1ea41e_48062"
    />
    <input
      autocomplete="off"
      id="email_subscribe_a7fdfc1ea41e_48062"
      name="subscribe_a7fdfc1ea41e_48062"
      placeholder="Your email here"
      tabindex="-1"
      type="email"
      value=""
    />
  </div>
  <input
    autocomplete="off"
    id="g-recaptcha-response"
    name="g-recaptcha-response"
    type="hidden"
    value=""
  />
  <button
    class="btn btn-primary"
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`ContactForm > should not submit when disabled > dirty 1`] = `
<form
  accept-charset="UTF-8"
  action="/submit"
  class="was-validated"
  data-testid="contact-form"
  enctype=""
  id="contact-form"
  method="GET"
  novalidate=""
>
  <div>
    <label
      class="form-label"
      for="contact.Full_Name"
    >
      Full Name *
    </label>
    <input
      class="form-control"
      id="contact.Full_Name"
      name="Full Name"
      placeholder="First name"
      required=""
      type="text"
    />
    <div
      class="invalid-feedback"
    >
      Please enter your full name
    </div>
  </div>
  <div>
    <label
      class="form-label"
      for="contact.Email_Address"
    >
      Email Address *
    </label>
    <input
      class="form-control"
      id="contact.Email_Address"
      name="Email Address"
      required=""
      type="email"
    />
    <div
      class="invalid-feedback"
    >
      Please enter a valid email address
    </div>
  </div>
  <div>
    <label
      class="form-label"
      for="contact.Comments_Questions"
    >
      Comments/Questions *
    </label>
    <textarea
      class="form-control"
      id="contact.Comments_Questions"
      name="Comments/Questions"
      required=""
    />
    <div
      class="invalid-feedback"
    >
      Please enter your feedback.
    </div>
  </div>
  <div
    style="opacity: 0; position: absolute; top: 0px; left: -5000px; height: 0px; width: 0px;"
  >
    <label
      for="subscribe_a7fdfc1ea41e_48062"
    />
    <input
      autocomplete="off"
      id="email_subscribe_a7fdfc1ea41e_48062"
      name="subscribe_a7fdfc1ea41e_48062"
      placeholder="Your email here"
      tabindex="-1"
      type="email"
      value=""
    />
  </div>
  <input
    autocomplete="off"
    id="g-recaptcha-response"
    name="g-recaptcha-response"
    type="hidden"
    value=""
  />
  <button
    class="btn btn-primary"
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`ContactForm > should not submit when disabled > pristine 1`] = `
<form
  accept-charset="UTF-8"
  action="/submit"
  class=""
  data-testid="contact-form"
  enctype=""
  id="contact-form"
  method="GET"
  novalidate=""
>
  <div>
    <label
      class="form-label"
      for="contact.Full_Name"
    >
      Full Name *
    </label>
    <input
      class="form-control"
      disabled=""
      id="contact.Full_Name"
      name="Full Name"
      placeholder="First name"
      required=""
      type="text"
    />
    <div
      class="invalid-feedback"
    >
      Please enter your full name
    </div>
  </div>
  <div>
    <label
      class="form-label"
      for="contact.Email_Address"
    >
      Email Address *
    </label>
    <input
      class="form-control"
      disabled=""
      id="contact.Email_Address"
      name="Email Address"
      required=""
      type="email"
    />
    <div
      class="invalid-feedback"
    >
      Please enter a valid email address
    </div>
  </div>
  <div>
    <label
      class="form-label"
      for="contact.Comments_Questions"
    >
      Comments/Questions *
    </label>
    <textarea
      class="form-control"
      disabled=""
      id="contact.Comments_Questions"
      name="Comments/Questions"
      required=""
    />
    <div
      class="invalid-feedback"
    >
      Please enter your feedback.
    </div>
  </div>
  <div
    style="opacity: 0; position: absolute; top: 0px; left: -5000px; height: 0px; width: 0px;"
  >
    <label
      for="subscribe_a7fdfc1ea41e_48062"
    />
    <input
      autocomplete="off"
      disabled=""
      id="email_subscribe_a7fdfc1ea41e_48062"
      name="subscribe_a7fdfc1ea41e_48062"
      placeholder="Your email here"
      tabindex="-1"
      type="email"
      value=""
    />
  </div>
  <input
    autocomplete="off"
    disabled=""
    id="g-recaptcha-response"
    name="g-recaptcha-response"
    type="hidden"
    value=""
  />
  <button
    class="btn btn-primary"
    disabled=""
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`ContactForm > should not submit when form has invalid data > dirty 1`] = `
<form
  accept-charset="UTF-8"
  action="/submit"
  class="was-validated"
  data-testid="contact-form"
  enctype=""
  id="contact-form"
  method="GET"
  novalidate=""
>
  <div>
    <label
      class="form-label"
      for="contact.Full_Name"
    >
      Full Name *
    </label>
    <input
      class="form-control"
      id="contact.Full_Name"
      name="Full Name"
      placeholder="First name"
      required=""
      type="text"
    />
    <div
      class="invalid-feedback"
    >
      Please enter your full name
    </div>
  </div>
  <div>
    <label
      class="form-label"
      for="contact.Email_Address"
    >
      Email Address *
    </label>
    <input
      class="form-control"
      id="contact.Email_Address"
      name="Email Address"
      required=""
      type="email"
    />
    <div
      class="invalid-feedback"
    >
      Please enter a valid email address
    </div>
  </div>
  <div>
    <label
      class="form-label"
      for="contact.Comments_Questions"
    >
      Comments/Questions *
    </label>
    <textarea
      class="form-control"
      id="contact.Comments_Questions"
      name="Comments/Questions"
      required=""
    />
    <div
      class="invalid-feedback"
    >
      Please enter your feedback.
    </div>
  </div>
  <div
    style="opacity: 0; position: absolute; top: 0px; left: -5000px; height: 0px; width: 0px;"
  >
    <label
      for="subscribe_a7fdfc1ea41e_48062"
    />
    <input
      autocomplete="off"
      id="email_subscribe_a7fdfc1ea41e_48062"
      name="subscribe_a7fdfc1ea41e_48062"
      placeholder="Your email here"
      tabindex="-1"
      type="email"
      value=""
    />
  </div>
  <input
    autocomplete="off"
    id="g-recaptcha-response"
    name="g-recaptcha-response"
    type="hidden"
    value=""
  />
  <button
    class="btn btn-primary"
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`ContactForm > should not submit when form has invalid data > pristine 1`] = `
<form
  accept-charset="UTF-8"
  action="/submit"
  class=""
  data-testid="contact-form"
  enctype=""
  id="contact-form"
  method="GET"
  novalidate=""
>
  <div>
    <label
      class="form-label"
      for="contact.Full_Name"
    >
      Full Name *
    </label>
    <input
      class="form-control"
      id="contact.Full_Name"
      name="Full Name"
      placeholder="First name"
      required=""
      type="text"
    />
    <div
      class="invalid-feedback"
    >
      Please enter your full name
    </div>
  </div>
  <div>
    <label
      class="form-label"
      for="contact.Email_Address"
    >
      Email Address *
    </label>
    <input
      class="form-control"
      id="contact.Email_Address"
      name="Email Address"
      required=""
      type="email"
    />
    <div
      class="invalid-feedback"
    >
      Please enter a valid email address
    </div>
  </div>
  <div>
    <label
      class="form-label"
      for="contact.Comments_Questions"
    >
      Comments/Questions *
    </label>
    <textarea
      class="form-control"
      id="contact.Comments_Questions"
      name="Comments/Questions"
      required=""
    />
    <div
      class="invalid-feedback"
    >
      Please enter your feedback.
    </div>
  </div>
  <div
    style="opacity: 0; position: absolute; top: 0px; left: -5000px; height: 0px; width: 0px;"
  >
    <label
      for="subscribe_a7fdfc1ea41e_48062"
    />
    <input
      autocomplete="off"
      id="email_subscribe_a7fdfc1ea41e_48062"
      name="subscribe_a7fdfc1ea41e_48062"
      placeholder="Your email here"
      tabindex="-1"
      type="email"
      value=""
    />
  </div>
  <input
    autocomplete="off"
    id="g-recaptcha-response"
    name="g-recaptcha-response"
    type="hidden"
    value=""
  />
  <button
    class="btn btn-primary"
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`ContactForm > should submit when form has valid data > dirty 1`] = `
<form
  accept-charset="UTF-8"
  action="/submit"
  class="was-validated"
  data-testid="contact-form"
  enctype=""
  id="contact-form"
  method="GET"
  novalidate=""
>
  <div>
    <label
      class="form-label"
      for="contact.Full_Name"
    >
      Full Name *
    </label>
    <input
      class="form-control"
      id="contact.Full_Name"
      name="Full Name"
      placeholder="First name"
      required=""
      type="text"
    />
    <div
      class="invalid-feedback"
    >
      Please enter your full name
    </div>
  </div>
  <div>
    <label
      class="form-label"
      for="contact.Email_Address"
    >
      Email Address *
    </label>
    <input
      class="form-control"
      id="contact.Email_Address"
      name="Email Address"
      required=""
      type="email"
    />
    <div
      class="invalid-feedback"
    >
      Please enter a valid email address
    </div>
  </div>
  <div>
    <label
      class="form-label"
      for="contact.Comments_Questions"
    >
      Comments/Questions *
    </label>
    <textarea
      class="form-control"
      id="contact.Comments_Questions"
      name="Comments/Questions"
      required=""
    />
    <div
      class="invalid-feedback"
    >
      Please enter your feedback.
    </div>
  </div>
  <div
    style="opacity: 0; position: absolute; top: 0px; left: -5000px; height: 0px; width: 0px;"
  >
    <label
      for="subscribe_a7fdfc1ea41e_48062"
    />
    <input
      autocomplete="off"
      id="email_subscribe_a7fdfc1ea41e_48062"
      name="subscribe_a7fdfc1ea41e_48062"
      placeholder="Your email here"
      tabindex="-1"
      type="email"
      value=""
    />
  </div>
  <input
    autocomplete="off"
    id="g-recaptcha-response"
    name="g-recaptcha-response"
    type="hidden"
    value=""
  />
  <button
    class="btn btn-primary"
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`ContactForm > should submit when form has valid data > pristine 1`] = `
<form
  accept-charset="UTF-8"
  action="/submit"
  class="was-validated"
  data-testid="contact-form"
  enctype=""
  id="contact-form"
  method="GET"
  novalidate=""
>
  <div>
    <label
      class="form-label"
      for="contact.Full_Name"
    >
      Full Name *
    </label>
    <input
      class="form-control"
      id="contact.Full_Name"
      name="Full Name"
      placeholder="First name"
      required=""
      type="text"
    />
    <div
      class="invalid-feedback"
    >
      Please enter your full name
    </div>
  </div>
  <div>
    <label
      class="form-label"
      for="contact.Email_Address"
    >
      Email Address *
    </label>
    <input
      class="form-control"
      id="contact.Email_Address"
      name="Email Address"
      required=""
      type="email"
    />
    <div
      class="invalid-feedback"
    >
      Please enter a valid email address
    </div>
  </div>
  <div>
    <label
      class="form-label"
      for="contact.Comments_Questions"
    >
      Comments/Questions *
    </label>
    <textarea
      class="form-control"
      id="contact.Comments_Questions"
      name="Comments/Questions"
      required=""
    />
    <div
      class="invalid-feedback"
    >
      Please enter your feedback.
    </div>
  </div>
  <div
    style="opacity: 0; position: absolute; top: 0px; left: -5000px; height: 0px; width: 0px;"
  >
    <label
      for="subscribe_a7fdfc1ea41e_48062"
    />
    <input
      autocomplete="off"
      id="email_subscribe_a7fdfc1ea41e_48062"
      name="subscribe_a7fdfc1ea41e_48062"
      placeholder="Your email here"
      tabindex="-1"
      type="email"
      value=""
    />
  </div>
  <input
    autocomplete="off"
    id="g-recaptcha-response"
    name="g-recaptcha-response"
    type="hidden"
    value=""
  />
  <button
    class="btn btn-primary"
    type="submit"
  >
    Submit
  </button>
</form>
`;
